src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,538 @@
+            }
+            /* else if (listener.SupportsRequestReply && typeof(TChannel) == typeof(IReplySessionChannel))
+             {
+                 securityChannel = new SecurityReplySessionChannel(listener, (IReplySessionChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+             }
+             else
+             {
+                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.GetString(SR.UnsupportedChannelInterfaceType, typeof(TChannel))));
+             }*/
+
+            return securityChannelDispatcher;
+        }
+    }
+
+    internal abstract class ServerSecurityChannel<UChannel> where UChannel : class //, IChannel
+    {
+        private static MessageFault secureConversationCloseNotSupportedFault;
+        private string secureConversationCloseAction;
+        private SecurityListenerSettingsLifetimeManager settingsLifetimeManager;
+        private bool hasSecurityStateReference;
+        
+        protected ServerSecurityChannel(SecurityServiceDispatcher securityServiceDispatcher, UChannel innerChannel, SecurityProtocol securityProtocol, SecurityListenerSettingsLifetimeManager settingsLifetimeManager)
+        // : base(channelManager, innerChannel, securityProtocol)
+        {
+            if (settingsLifetimeManager == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("settingsLifetimeManager");
+            }
+            this.settingsLifetimeManager = settingsLifetimeManager;
+            this.SecurityProtocol = securityProtocol;
+            this.OuterChannel = (IReplyChannel)innerChannel;
+        }
+
+        internal SecurityProtocol SecurityProtocol { get; set; }
+        public IReplyChannel OuterChannel { get; private set; }
+
+        private static MessageFault GetSecureConversationCloseNotSupportedFault()
+        {
+            if (secureConversationCloseNotSupportedFault == null)
+            {
+                FaultCode faultCode = FaultCode.CreateSenderFaultCode(DotNetSecurityStrings.SecureConversationCancelNotAllowedFault, DotNetSecurityStrings.Namespace);
+                FaultReason faultReason = new FaultReason(SR.Format(SR.SecureConversationCancelNotAllowedFaultReason), System.Globalization.CultureInfo.InvariantCulture);
+                secureConversationCloseNotSupportedFault = MessageFault.CreateFault(faultCode, faultReason);
+            }
+            return secureConversationCloseNotSupportedFault;
+        }
+
+        private void ThrowIfSecureConversationCloseMessage(Message message)
+        {
+            if (message.Headers.Action == this.secureConversationCloseAction)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new MessageSecurityException(SR.Format(SR.SecureConversationCancelNotAllowedFaultReason), null, GetSecureConversationCloseNotSupportedFault()));
+            }
+        }
+
+        [SecuritySafeCritical]
```

Comment:
This attribute is a remnant of the CAS feature in .NET Framework which was deprecated long ago. If you ever see a `[SeuritySafeCritical]` attribute applied somewhere, just remove it.